Service Crate Generator

mthjones opened the pull request:
Here it is. This builds upon the work done in #623 -- the diff will shrink when that is merged. Sorry that it's so large, I tried to find ways to shrink the code that needed review, but it's such a large refactoring that I don't think it's possible to get it much smaller than this.

I've labelled this as a WIP, but it should have feature parity with the existing `rusoto` code. Remaining work is around code cleanup and improving crate generator documentation. Also making sure that Travis/Appveyor work, which I can only do by opening a PR.

## Motivation
The rusoto crate suffers from ridiculously long compile times, causing CI timeouts and slow builds for users. In an attempt to alleviate this compile time, it was suggested to split the project into individual crates that can be built and tested separately.

## Description
This PR reworks how services are generated by implementing a new executable project `services_crategen` that can generate individual crates for each service. The executable takes two inputs:
* the list of services to generate, either directly on the command line in `name@version` format or stored in a configuration file. The standard config file is provided as `services.json` in `service_crategen`.
* the output folder to generate the crates in. This should almost always be `../rusoto/services` unless testing.

The executable can be rerun and will produce identical output, provided nothing in the service definition or the code generation has changed.

Once the service's crate has been generated, it can be customized. Certain crates contain custom tests, and `sts` contains custom credential implementation logic. This custom logic can be placed in the `custom` module in a crate, which will not be overwritten on crate regeneration (but obviously must be kept up-to-date in order to ensure the crate continues to build).

Building and testing the full project is nearly as simple as it was before, using Cargo workspaces. Just go to the `rusoto` directory and `cargo build --all` or `cargo test --all`. Only this time it shouldn't bring your machine to its knees. 😁

**Note**: This PR does _not_ contain the generated service code. That will come in a future PR, pending discussion in #559.

## Future Enhancements
* ~~Generate README and LICENSE along with each crate~~
* Move integration tests into each service crate
* Improve modularization of crate generator code. Currently just an MVP
* Remove "meta-crate"? This is the current `rusoto` crate with features for services, implemented using the crates and core module instead.
* Expand declarative service configuration (`services.json`) to include things like custom dependencies.

adimarco commented:
I can't tell you how happy I am about this.

mthjones commented:
Looking into Travis failure (looks like `travis-cargo` issues), but appveyor succeeded so we have some CI numbers to go off of here:

[Build #552](https://ci.appveyor.com/project/matthewkmayer/rusoto/build/1.0.552):
`stable`: 21:42
`1.16.0`: 18:53

[Build #560](https://ci.appveyor.com/project/matthewkmayer/rusoto/build/1.0.560): (this one)
`stable`: 12:40
`1.16.0`: 11:58

So about a 1/3 time reduction. Note that this also includes generating the service crates as well, which will be going away, based on recent comments in #559. Also doing build and test in two separate steps, which is unnecessarily doubling the time, but for a fair comparison I left that as-is.

adimarco commented:
I can confirm that the crategen works for me, and that the build and tests all work as well.  And quickly.

I'm giddy at the prospect of getting this merged.  I'll read over the code tomorrow and fast-track this whole thing.

I think the biggest hurdle left here is going to be figuring out what the impact the the end-user of the rusoto crate is, and just getting all our documentation up to date.

mthjones commented:
Interesting... Travis finished successfully, but the times are actually not much faster on the Linux builds. However, the OSX builds are running about 40-50% faster. It comes out to about a 20-40min overall reduction, judging from recent CI builds. However, it also does give the advantage of not having future service additions slowing down the build even more significantly, and no more need to fiddle with `travis_wait`.

adimarco commented:
Looks like there's maybe an extra "rusoto" in there:

```
Cargo failed to read `--manifest-path /home/travis/build/rusoto/rusoto/rusoto/Cargo.toml` and `--manifest-path /home/travis/build/rusoto/rusoto/rusoto`
```

mthjones commented:
Heh, as silly as that looks, it's actually correct. It checks out the project as `rusoto/rusoto` (organization + repository name from Github), then uses the `rusoto` folder inside.

The issue is in the doc generation build step. `travis-cargo` doesn't handle passing `--all` it seems like, so I think we'll have to use regular cargo and figure out a different way to do `--only stable`.

mthjones commented:
Regarding this:

> I think the biggest hurdle left here is going to be figuring out what the impact to the end-user of the rusoto crate is from this change, and just getting all our documentation up to date.

One thing to think about is what to do with the `rusoto` crate name. With `rusoto_core` now, and all of these services in separate crates, what does the `rusoto` crate mean? Should it exist anymore or should it be deprecated in favor of using service crates?

adimarco commented:
I think it might just end up being a "brand" name for the overall project, and maybe the domain name where the documentation lives?  I think we might want to open up an issue to discuss that.

mthjones commented:
Pushed a new commit that should fix the Travis issue, also rebased off master to resolve conflicts.

Let me know if it's too much to review. I realized today that I could have left the "metacrate" work for another PR, which would shrink the diff to review in this one.

adimarco commented on Cargo.toml:
```diff
@@ -21,35 +20,294 @@ homepage = "https://www.rusoto.org/"
-[dependencies.rusoto_core]
-path = "rusoto/core"
+[dependencies.rusoto_cloudformation]
+optional = true
```
D'oh.  This little detail was a big piece of what I was missing conceptually when I tried to get this split out but still using the same workspace.  It never even occurred to me to declare the sub-crate dependencies optional.  Turns out that NOT doing that makes this mostly unworkable.

adimarco commented on service_crategen/services.json:
```diff
@@ -0,0 +1,170 @@
+{
```
Thought for the future - this default config could all be generated by walking the botocore services directory and parsing out the most recent date.

mthjones replied:
Perhaps. If it just scans the botocore directory, we lose control over which services to generate. This is sort of based off [this config](https://github.com/aws/aws-sdk-net/blob/master/generator/ServiceModels/_sdk-versions.json) in the .Net SDK. I think it could be expanded in the future though. One thing I'd probably like to do is decouple the versions from each other.

adimarco commented on service_crategen/src/cargo.rs:
```diff
@@ -0,0 +1,52 @@
+use std::collections::BTreeMap;
+
+#[derive(Debug, Default, Clone, Deserialize, Serialize)]
+pub struct Manifest {
```
I'm surprised this stuff doesn't exist in some already-available crate?  Surely we're not the only ones to build a cargo manifest programmatically.

mthjones replied:
I know! I couldn't find anything to use for this. I figured maybe the `cargo` crate would, but it doesn't. `cargo-edit`, also does not, it works on `toml::Value`. If anything comes up, this should be pretty easy to swap out for something else.

adimarco commented on service_crategen/src/main.rs:
```diff
@@ -0,0 +1,287 @@
+
+    match &service.metadata.protocol[..] {
+        "json" => {
+            dependencies.insert("serde".to_owned(), cargo::Dependency::Simple("0.9.5".into()));
```
I do like the idea of building this cargo manifest with code.  I don't like keeping these lists of dependencies and their versions as "magic numbers" in the code.  Making a rust code change to update a `serde` dependency version seems counter-intuitive, even if it's a dependency in a crate generated by this executable.  

Not a reason to stop this merge, but we should open an issue for making this code read some kind of configuration file and do the exact same thing it's doing now with that configuration.

mthjones replied:
Agreed. I initially looked at using template files for these but couldn't find a template library I was satisfied with. Can definitely be a future enhancement though.

For the dependency numbers, that one's a bit complicated. Should versions be globally set or defined per service? Should all services require updating if you update the dependency for one of them? Also, some services have custom dependencies, like `s3`.

adimarco commented on service_crategen/src/main.rs:
```diff
@@ -0,0 +1,287 @@
+            .expect("Unable to write Cargo.toml");
+        
+        let manifest = cargo::Manifest {
+            package: cargo::Metadata {
```
As with the above comment, maybe this preamble should all be in a `.toml` file somewhere that gets read in and either just straight regurgitated back out, or parsed and manipulated if needs be.  It'd make tweaking it a bit easier.

adimarco commented on src/lib.rs:
```diff
@@ -45,149 +45,124 @@
-
 #[cfg(feature = "acm")]
-pub mod acm;
+pub extern crate rusoto_acm as acm;
```
This gets us into the larger discussion of "how will these crates all get packaged once this is landed", but that's a discussion for another PR.

mthjones replied:
Yup. With merging this PR as it is, the `rusoto` crate could still be published and used as-is (though with better compilation times).

adimarco commented:
Complicating things a wee bit with the dependencies for this - I just merged #625 which updates the serde dependency to 1.0.

mthjones commented:
Rebased on the serde 1.0 update.

adimarco commented:
Do you want to get this merged in before #620 or vice-versa?  That's another PR I'd like to get merged for the next release while we're breaking everything anyway.

mthjones commented:
Either way works for me. Probably easiest to update this PR after #620 gets merged than the other way around though.

mrhota commented on service_crategen/codegen/src/botocore.rs:
```diff
@@ -1,7 +1,15 @@
 
 use serialization::{ShapesMap, ShapeName};
 
+const BOTOCORE_DIR: &'static str = concat!(env!("CARGO_MANIFEST_DIR"), "/botocore/botocore/data/");
```
will this work on non-*nix?

matthewkmayer commented:
I tried following what the travis build file does by doing a `cargo clean` in the root project directory then `cargo run -- -c ./services.json -o ../rusoto/services` in the `service_crategen` directory.  I saw a lot of errors like this:

`Failed to load service sns: No such file or directory (os error 2)`

Am I missing a setup step?

mthjones commented:
Make sure to run `git submodule update --init` first. :)

I'll add a basic README to the service_crategen project to document how to run it more clearly.

mthjones commented:
README added: https://github.com/mthjones/rusoto/blob/1c0c1cdc4ac888da3eab0e885d9f0b8be434ce4f/service_crategen/README.md

Hope that helps. :)

matthewkmayer commented:
Thanks, that cleared it up and crategen works as expected! 🎉 

After reviewing the code, my only concern is API docs.  I ran what Travis runs (`cd rusoto && cargo doc --all --no-deps`) and while it does generate docs for all the service crates, they don't show up on the list of crates when opening up the docs.  I only see rusoto and rusoto_credential.  Maybe I'm looking at the wrong output dir.

If I run `cargo doc --all` and open something such as `rusoto/rusoto/target/doc/rusoto_cloudformation/index.html` I see all the rusoto crates, but also our dependencies.

Ideally we'd have only the rusoto crates documented so it's easier to find things.  Perhaps a meta-crate for docs is a solution.  Hopefully I'm missing something. 😄 

Once we get the API docs figured out this is ready to merge.  Thanks for the hard work on this!

mthjones commented:
Hmm, that's strange. If I run `cargo doc --all --nodeps` and open `rusoto/rusoto/target/doc/rusoto_cloudformation/index.html`, I get the list of crates with no dependencies.

![crates](https://cloud.githubusercontent.com/assets/45536/25775295/22687678-3270-11e7-904f-f33ce3589048.PNG)

Are you running that command from the repository root or the `rusoto` directory inside it? From the root, there's still the "meta-crate" there, which still requires features to include services. I'm not sure why it even accepts `--all` as a valid parameter since it's not a workspace...

However, I don't know how travis-cargo doc-upload works with a workspace. My guess is that it just doesn't, so we'd have to use the "meta-crate" for now.

Another option is to just rely on docs.rs. Since they will all now be published crates, their documentation will all be picked up and hosted. The website could contain a directory of links to each service's crate documentation, with each crate's README having a link back to the directory.

mthjones commented:
Ehh, the metacrate documentation doesn't work with `--no-deps`, since each crate is a dependency. Hrm...

mthjones commented:
Pushed a change to generate the README along with each service crate. Fairly simple right now (pretty much all just taken from the top-level README), but it should make it easier to add some documentation to the crates.

matthewkmayer commented:
I ran `cd rusoto && cargo doc --all --no-deps` on my Linux desktop and it works as expected and how your screenshot shows it.  I wonder if my laptop has some leftover files from running that command on master.  I should try a fresh cloning of the repo and check this PR again.

One thing I did see is we lost ease of finding the Client from https://github.com/rusoto/rusoto/issues/519 .  The trait-based setup we now use is easier to find but I'd love to call it out more.  The need for this is apparent when you open the docs for `rusoto_s3` and try to find out how to use the crate. :smile:

mthjones commented:
Added the top-level docs back for each service. One thing I think would be nice is to generate all request/response types into a separate module, so the top-level just contains the client and trait with the types module publicly exported. But for now the documentation works.

matthewkmayer commented:
I apparently had some leftover directories from my branch for https://github.com/rusoto/rusoto/pull/600 that git wasn't mentioning.  Got the docs to work as expected and checked the trait and client links, working great now!

The `rusoto_credential` crate appears to be missing from documentation though.

mthjones commented:
`rusoto_credential` was left in its original location and it will be handled the same way as it was before. `rusoto_core` re-exports most of `rusoto_credential`, just like the main `rusoto` crate did. I don't believe `rusoto_credential` was included in the main crate documentation.

mthjones commented:
Ah, I see. So the doc generation included rusoto_credential before. Ok, I can just move `credential` into the workspace.

mthjones commented:
Moved `credential` into the workspace so it's built, tested, and documented alongside the rest of the published crates. Also moved `mock` out, as it's just an internal testing helper crate and should not be getting documented with the published crates.

matthewkmayer commented:
This PR:

* has crategen working
* API docs generated
* passing CI builds
* (most) integration tests pass

The elastictranscoder test needs the S3 trait imported in order to run its integration tests.  I think that's the last thing for us to address in this PR. :tada:

mthjones commented:
Done. Also updated the `services.json` configuration to be more extendable and rely on less baked-in customization in the code itself.

mthjones commented:
Oh, one question before this gets merged. The `cognito-identity` service has a dash in the name, unlike any other AWS service. This _works_ in a crate name, but is a bit unusual when paired with the `rusoto_` prefix. It turns into `rusoto_cognito-identity`. Should I convert dashes to underscores when generating the crates?

adimarco commented:
Yeah, let's translate all dashes to underscores.  I'm frankly surprised it does work in a crate name.  I've done a double-take over it before.

mthjones commented:
Done.

matthewkmayer commented:
Anything left to do on this?  The undocumented checklist I've been using is:

* ensure it compiles
* integration tests run and pass
* API documentation is created

mthjones commented:
Nope, unless there's any more feedback!
On Wed, May 10, 2017 at 1:29 AM Matthew Mayer <notifications@github.com>
wrote:

> Anything left to do on this? The undocumented checklist I've been using is:
>
>    - ensure it compiles
>    - integration tests run and pass
>    - API documentation is created
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rusoto/rusoto/pull/628#issuecomment-300377993>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACx4BN1vhyn2iRKMbg4rcm16hFcN-H-ks5r4Us0gaJpZM4NMuct>
> .
>

matthewkmayer commented:
🎉

mthjones commented:
Nice, thanks! I think the only thing left is getting the generated code checked in now. Working on integrating rustfmt into the generation, but it throws errors in Powershell due to a bug in Syntex:
- https://github.com/rust-lang-nursery/rustfmt/issues/1180
- https://github.com/rust-lang-nursery/rustfmt/issues/1525

I could check it in without rustfmt integration, but there would be another large churn PR if rustfmt integration were added at a later date.

mthjones commented:
Also, removing the `rusoto` crate. What did we want to do about that?

mthjones commented:
Ahhh, and doc upload failed. Argh. Will take a look, travis-cargo might be unworkable.
